backends/p4fmt/p4fmt.cpp

```diff
@@ -18,13 +56,29 @@ std::stringstream getFormattedOutput(std::filesystem::path inputFile) {
 
     std::stringstream formattedOutput;
 
-    const IR::P4Program *program = P4::parseP4File(options);
+    auto result = parseProgram(options);
+
+    const IR::P4Program *program = result.first;
+    const Util::InputSources *sources = result.second;
+
     if (program == nullptr && ::P4::errorCount() != 0) {
         ::P4::error("Failed to parse P4 file.");
         return formattedOutput;
     }
 
+    std::unordered_map<const Util::Comment *, bool> globalCommentsMap;
+
+    // Initialize the global comments map from the list of comments in the program.
+    if (sources != nullptr) {
+        for (const auto *comment : sources->getAllComments()) {
+            globalCommentsMap[comment] =
```

Comment:
This makes the assumption that comment pointers from different sources are the same, right? Otherwise, we may get duplicates here.